src/3d/qgsvirtualpointcloudentity_p.cpp

```diff
@@ -0,0 +1,201 @@
+    float maximumScreenSpaceError,
+    bool showBoundingBoxes,
+    double zValueScale,
+    double zValueOffset,
+    int pointBudget )
+  : Qgs3DMapSceneEntity( nullptr )
+  , mLayer( layer )
+  , mMap( map )
+  , mCoordinateTransform( coordinateTransform )
+  , mZValueScale( zValueScale )
+  , mZValueOffset( zValueOffset )
+  , mPointBudget( pointBudget )
+  , mMaximumScreenSpaceError( maximumScreenSpaceError )
+  , mShowBoundingBoxes( showBoundingBoxes )
+{
+  mSymbol.reset( symbol );
+  mBboxesEntity = new QgsChunkBoundsEntity( this );
+  const QgsRectangle mapExtent = Qgs3DUtils::tryReprojectExtent2D( mMap.extent(), mMap.crs(), layer->crs(), mMap.transformContext() );
+  const QVector<QgsPointCloudSubIndex> subIndexes = provider()->subIndexes();
+  for ( int i = 0; i < subIndexes.size(); ++i )
+  {
+    const QgsPointCloudSubIndex &si = subIndexes.at( i );
+    const QgsRectangle intersection = si.extent().intersect( mapExtent );
+
+    mBboxes << Qgs3DUtils::mapToWorldExtent( intersection, si.zRange().lower(), si.zRange().upper(), mMap.origin() );
+
+    if ( !si.index() || intersection.isEmpty() )
+      continue;
+
+    QgsPointCloudLayerChunkedEntity *entity = new QgsPointCloudLayerChunkedEntity( si.index(),
+        mMap,
+        mCoordinateTransform,
+        static_cast< QgsPointCloud3DSymbol * >( mSymbol->clone() ),
+        mMaximumScreenSpaceError,
+        mShowBoundingBoxes,
+        mZValueScale,
+        mZValueOffset,
+        mPointBudget );
+
+    mChunkedEntitiesMap.insert( i, entity );
+    entity->setParent( this );
+    connect( entity, &QgsChunkedEntity::pendingJobsCountChanged, this, &Qgs3DMapSceneEntity::pendingJobsCountChanged );
+  }
+
+  updateBboxEntity();
+  connect( provider(), &QgsVirtualPointCloudProvider::subIndexLoaded, this, &QgsVirtualPointCloudEntity::createChunkedEntityForSubIndex );
+}
+
+QList<QgsChunkedEntity *> QgsVirtualPointCloudEntity::chunkedEntities() const
+{
+  return mChunkedEntitiesMap.values();
+}
+
+QgsVirtualPointCloudProvider *QgsVirtualPointCloudEntity::provider() const
+{
+  return qobject_cast<QgsVirtualPointCloudProvider *>( mLayer->dataProvider() );
+}
+
+QgsAABB QgsVirtualPointCloudEntity::boundingBox( int i ) const
+{
+  return mBboxes.at( i );
+}
+
+void QgsVirtualPointCloudEntity::createChunkedEntityForSubIndex( int i )
+{
+  const QVector<QgsPointCloudSubIndex> subIndexes = provider()->subIndexes();
+  QgsPointCloudLayerChunkedEntity *newChunkedEntity = new QgsPointCloudLayerChunkedEntity( subIndexes.at( i ).index(),
+      mMap,
+      mCoordinateTransform,
+      static_cast< QgsPointCloud3DSymbol * >( mSymbol->clone() ),
+      mMaximumScreenSpaceError,
+      mShowBoundingBoxes,
+      mZValueScale,
+      mZValueOffset,
+      mPointBudget );
+
+  mChunkedEntitiesMap.insert( i, newChunkedEntity );
+  newChunkedEntity->setParent( this );
+  connect( newChunkedEntity, &QgsChunkedEntity::pendingJobsCountChanged, this, &Qgs3DMapSceneEntity::pendingJobsCountChanged );
```

Comment:
It would be good to de-duplicate the code here and in the constructor that creates child chunked entities. There seems to be a slight difference in behavior as well compared to the for-loop in the constructor - e.g. here we are not doing 3d map extent intersection test (so if I understand correctly, if 2D canvas loads a sub-index, it will still show up in 3d map even if it is outside)

Reply:
Hmm, good point, let me simplify this.